### PR TITLE
Focal Operations and Buffer Tile support

### LINKED ISSUE #441: Implement buffered tiles

### PR DESCRIPTION
**Pull request by pomadchin**
It is a port of https://github.com/locationtech/rasterframes/pull/551

Known limitations:

- [x] Focal Expressions should be of a correct arity
- [x] Python Focal API ~, blocked by https://github.com/locationtech/rasterframes/pull/562~
  - Docker image for tests: `quay.io/daunnc/rasterframes-notebook:0.9.2-SNAPSHOT`
- [x] Python STAC API
- [x] After reading a buffered tile and performing a local operation the buffer is lost: I propose that to be not a part of this PR, but to move it into a separate one as an enchancement. 

#### Focal Operations notebook screenshot
image

#### STAC API Example notebook screenshot
image

Closes #441 
Closes #551 
Closes #487

Blocked by https://github.com/locationtech/geotrellis/pull/3419

### CODE REVIEW

File: core/src/main/scala/org/locationtech/rasterframes/expressions/focalops/SurfaceOp.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import geotrellis.raster.Tile
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.{TypeCheckFailure, TypeCheckSuccess}
+import org.apache.spark.sql.catalyst.expressions.UnaryExpression
+import org.locationtech.rasterframes.model.TileContext
+import org.locationtech.rasterframes.expressions.NullToValue
+import org.locationtech.rasterframes.tiles.ProjectedRasterTile
+
+/** Operation on a tile returning a tile. */
+trait SurfaceOp extends UnaryExpression with NullToValue with CodegenFallback  {
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/expressions/focalops/SurfaceOp.scala:42]: How is this different from [`UnaryRasterOp`](https://github.com/locationtech/rasterframes/blob/develop/core/src/main/scala/org/locationtech/rasterframes/expressions/UnaryRasterOp.scala)

File: datasource/src/test/scala/examples/BufferTiles.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      // .select(rf_hillshade($"proj_raster", 315, 45, 1))
+      .as[Option[ProjectedRasterTile]]
+      // .show(false)
+      .first()
+
+  // tile.get.renderPng().write("/tmp/hillshade-buffered.png")
+  // tile.get.renderPng().write("/tmp/hillshade-nobuffered.png")
+
+  // spark.stop()
+}
```
Comment (metasim) [inline datasource/src/test/scala/examples/BufferTiles.scala:68]: Wow....  this is fantastic work!!!

Comment (metasim) [reply datasource/src/test/scala/examples/BufferTiles.scala:68]: How did you do this without substantially changing `TileUDT`? Where is the buffer size encoded in the dataframe?

File: core/src/main/scala/org/locationtech/rasterframes/expressions/focalops/SurfaceOp.scala (hunk 1 above)
Comment (pomadchin) [reply core/src/main/scala/org/locationtech/rasterframes/expressions/focalops/SurfaceOp.scala:42]: @metasim _have no idea_; I also haven't renamed classes yet (talking about your changes in your PR).

File: datasource/src/test/scala/examples/BufferTiles.scala (hunk 2 above)
Comment (pomadchin) [reply datasource/src/test/scala/examples/BufferTiles.scala:68]: @metasim the it is a part of the RasterRef now, only raster refs for now can contain bufferSize; RasterRef is a case class.

Comment (pomadchin) [reply datasource/src/test/scala/examples/BufferTiles.scala:68]: @metasim check how we can flatten inputs and extract the BufferTile https://github.com/locationtech/rasterframes/pull/563/files#diff-f3a16be8b11d3ca1fd027d55f7ab7e9f06e4ccb312ea0e4760d1361016ae9fa5R57-R66 Later it is possible to pattern match on it. 

RasterRef operates by default with BufferTiles, the default buffer size = 0.

Comment (pomadchin) [reply datasource/src/test/scala/examples/BufferTiles.scala:68]: To improve the seamless integraion we can add the buffer size into the TileUDT tbh, that can extend all ops to work with the explicitly BufferTile object passed into the DataFrame.

File: core/src/main/scala/geotrellis/raster/BufferTile.scala [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+ *
+ * When combined with another BufferTile the operation will be executed over the maximum shared in
+ *
+ * TODO:
+ * - What should .map do? Map the buffer pixels or not?
+ * - toString method is friendly
+ * - mutable version makes sense
+ * - toBytes needs to encode padding size?
+ */
+case class BufferTile(
```
Comment (pomadchin) [inline core/src/main/scala/geotrellis/raster/BufferTile.scala:36]: TODO: merge into GeoTrellis

File: core/src/main/scala/org/locationtech/rasterframes/expressions/focalops/SurfaceOp.scala (hunk 1 above)
Comment (pomadchin) [reply core/src/main/scala/org/locationtech/rasterframes/expressions/focalops/SurfaceOp.scala]: @metasim restructured the hierarchy a bit and aligned with your changes from other PRs. I hope it looks a bit better now.

File: core/src/main/scala/org/apache/spark/sql/rf/TileUDT.scala [hunk 4]
```diff
@@ -59,34 +62,38 @@ class TileUDT extends UserDefinedType[Tile] {
...
+        InternalRow(ct, ref.cols, ref.rows, null, null, ref.toInternalRow)
       case prt: ProjectedRasterTile =>
         val tile = prt.tile
         val ct = UTF8String.fromString(tile.cellType.toString())
-        InternalRow(ct, tile.cols, tile.rows, tile.toBytes(), null)
+        InternalRow(ct, tile.cols, tile.rows, tile.toBytes(), null, null)
+      case bt: BufferTile =>
+        val tile = bt.toArrayTile()
+        val ct = UTF8String.fromString(tile.cellType.toString())
+        InternalRow(ct, tile.cols, tile.rows, tile.toBytes(), bt.gridBounds.toInternalRow, null)
```
Comment (pomadchin) [inline core/src/main/scala/org/apache/spark/sql/rf/TileUDT.scala:76]: I don't know how much that makes sense though, mb it is better not to preserve TileUDT "polymorphism" there. cc @metasim

File: core/src/main/scala/geotrellis/raster/BufferTile.scala (hunk 3 above)
Comment (pomadchin) [reply core/src/main/scala/geotrellis/raster/BufferTile.scala:36]: https://github.com/locationtech/geotrellis/pull/3419

File: core/src/main/scala/org/apache/spark/sql/rf/TileUDT.scala (hunk 4 above)
Comment (metasim) [reply core/src/main/scala/org/apache/spark/sql/rf/TileUDT.scala]: Tile polymorphism is really important for user ergnomics. Can all Tiles be BufferTiles?

File: datasource/src/test/scala/examples/BufferTiles.scala (hunk 2 above)
Comment (metasim) [reply datasource/src/test/scala/examples/BufferTiles.scala:68]: > extend all ops to work with the explicitly BufferTile object passed into the DataFrame

Can we do this without every Expression having to have this logic?

Comment (metasim) [reply datasource/src/test/scala/examples/BufferTiles.scala:68]: > @metasim check how we can flatten inputs and extract the BufferTile https://github.com/locationtech/rasterframes/pull/563/files#diff-f3a16be8b11d3ca1fd027d55f7ab7e9f06e4ccb312ea0e4760d1361016ae9fa5R57-R66 Later it is possible to pattern match on it.

Not sure this link is working still..


> RasterRef operates by default with BufferTiles, the default buffer size = 0.

And I'm wondering: what if all of RasterFrames worked this way?

**Comment by metasim** [comment]
@pomadchin I'm not sure I understand what you mean by these two sentenses:

> All functions can't be used in `spark-sql`. All Focal operations are unary and can't be applied to two expressions as that can be expected by the user.

**Comment by pomadchin** [comment]
@metasim Do we want to allow users to write cql expressions: 

```
select rf_hillshade(tile, 400, 10, 1)
```

Right now Focal ops are not allowing that, should not Hillshade ops in the case be `QuaternaryExpression` to parse all arguments?

File: core/src/main/scala/org/apache/spark/sql/rf/TileUDT.scala (hunk 4 above)
Comment (pomadchin) [reply core/src/main/scala/org/apache/spark/sql/rf/TileUDT.scala]: @metasim this change is pretty minor and keeps the PR within its scope: https://github.com/locationtech/rasterframes/pull/563/files#diff-4c1508a758314ff3b5773ada749119d0853a72490b4836749acf00b302fbc74dR59-R92 

both RasterRef and BufferTiles work as expected

**Comment by metasim** [comment]
> @metasim Do we want to allow users to write cql expressions:
> 
> ```
> select rf_hillshade(tile, 400, 10, 1)
> ```
> 
> Right now Focal ops are not allowing that, should not Hillshade ops in the case be `QuaternaryExpression` to parse all arguments?

Yes, eventually... I try to maintain Scala, Python and SQL parity. I'll look more deeply on what the issue is and what the workarounds might be, because I don't ever remember this being an issue (if you have the function as an Expression, it works in all the contexts).

File: core/src/test/scala/org/locationtech/rasterframes/RasterLayerSpec.scala [hunk 5]
```diff
@@ -76,6 +76,9 @@ class RasterLayerSpec extends TestEnvironment with MetadataKeys
   }
 
   describe("RasterFrameLayer") {
+    // Try to GC to avoid OOM on low memory instances.
+    // TODO: remove once we have a larger CI
+    System.gc()
```
Comment (metasim) [inline core/src/test/scala/org/locationtech/rasterframes/RasterLayerSpec.scala:81]: 🤣

File: core/src/main/scala/org/locationtech/rasterframes/expressions/UnaryRasterFunction.scala [hunk 6]
```diff
@@ -21,34 +21,27 @@
...
-    }
-    else TypeCheckSuccess
+    } else TypeCheckSuccess
   }
 
   override protected def nullSafeEval(input: Any): Any = {
-    val (childTile, childCtx) = tileExtractor(child.dataType)(row(input))
-    val result = op(childTile)
-    toInternalRow(result, childCtx)
+    // TODO: Ensure InternalRowTile is preserved
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/expressions/UnaryRasterFunction.scala:40]: This reminds me... if we want to start improving performance (mainly reducing copies), we need to figure out how ensure `InternalRowTile <op> R =>  InternalRowTile`. Either requires changes to how GeoTrellis instantiates destination tiles, or reimplementation of some fundamental operations.

File: core/src/main/scala/org/locationtech/rasterframes/functions/FocalFunctions.scala [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    Slope(tileCol, zFactorCol)
+
+  def rf_aspect(tileCol: Column): Column =
+    Aspect(tileCol)
+
+  def rf_hillshade[T: Numeric](tileCol: Column, azimuth: T, altitude: T, zFactor: T): Column =
+    rf_hillshade(tileCol, lit(azimuth), lit(altitude), lit(zFactor))
+
+  def rf_hillshade(tileCol: Column, azimuth: Column, altitude: Column, zFactor: Column): Column =
+    Hillshade(tileCol, azimuth, altitude, zFactor)
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/functions/FocalFunctions.scala:94]: 💯

File: core/src/main/scala/org/locationtech/rasterframes/ref/RasterRef.scala [hunk 8]
```diff
@@ -34,35 +34,49 @@ import org.locationtech.rasterframes.tiles.ProjectedRasterTile
...
+    // Pixel bounds we can read, including buffer
+    val possibleGrid = bufferedGrid.intersection(source.gridBounds).get
+    // Pixel bounds of center/non-buffer pixels in read tile
+    val tileCenterBounds = grid.offset(
+      colOffset = - possibleGrid.colMin,
+      rowOffset = - possibleGrid.rowMin
+    )
+
+    val raster = source.read(possibleGrid, Seq(bandIndex)).mapTile(_.band(0))
+    BufferTile(raster.tile, tileCenterBounds)
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/ref/RasterRef.scala:63]: 🙇

File: core/src/main/scala/org/locationtech/rasterframes/ref/RasterRef.scala [hunk 9]
```diff
@@ -34,35 +34,49 @@ import org.locationtech.rasterframes.tiles.ProjectedRasterTile
  *
  * @since 8/21/18
  */
-case class RasterRef(source: RFRasterSource, bandIndex: Int, subextent: Option[Extent], subgrid: Option[Subgrid]) extends ProjectedRasterTile {
+case class RasterRef(source: RFRasterSource, bandIndex: Int, subextent: Option[Extent], subgrid: Option[Subgrid], bufferSize: Short) extends ProjectedRasterTile {
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/ref/RasterRef.scala:37]: Why do you need both a `subgrid` and a `bufferSize`?

File: core/src/test/scala/org/locationtech/rasterframes/functions/FocalFunctionsSpec.scala [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+
+import java.nio.file.Paths
+
+class FocalFunctionsSpec extends TestEnvironment with RasterMatchers {
+
+  import spark.implicits._
+
+  describe("focal operations") {
+    lazy val path =
+      if(Paths.get("").toUri.toString.endsWith("core/")) Paths.get("src/test/resources/L8-B7-Elkton-VA.tiff").toUri
```
Comment (metasim) [inline core/src/test/scala/org/locationtech/rasterframes/functions/FocalFunctionsSpec.scala:42]: Magical dragons?

File: core/src/test/scala/org/locationtech/rasterframes/functions/FocalFunctionsSpec.scala [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+      // }
+
+      // that's what we have actually
+      // even though local ops can preserve the output tile
+      // we don't handle that
+      val expected = bt.localAdd(bt).aspect(btCellSize)
+      assertEqual(expected, actual)
+    }
+  }
+}
```
Comment (metasim) [inline core/src/test/scala/org/locationtech/rasterframes/functions/FocalFunctionsSpec.scala:346]: 💯

File: datasource/src/main/scala/org/locationtech/rasterframes/datasource/raster/RasterSourceDataSource.scala [hunk 12]
```diff
@@ -192,6 +197,11 @@ object RasterSourceDataSource {
...
       )
 
+    def withBufferSize(bufferSize: Short): TaggedReader =
+      tag[ReaderTag][DataFrameReader](
+        reader.option(RasterSourceDataSource.BUFFER_SIZE_PARAM, bufferSize)
+      )
+
     /** Indicate if tile reading should be delayed until cells are fetched. Defaults to `true`. */
     def withLazyTiles(state: Boolean): TaggedReader =
       tag[ReaderTag][DataFrameReader](
```
Comment (metasim) [inline datasource/src/main/scala/org/locationtech/rasterframes/datasource/raster/RasterSourceDataSource.scala:207]: Thanks for conforming to the pattern herein. All for the user ergonomics.

File: pyrasterframes/src/main/python/pyrasterframes/__init__.py [hunk 13]
```diff
@@ -134,6 +135,7 @@ def _raster_reader(
     :param catalog_col_names: required if `source` is a DataFrame or CSV string. It is a list of strings giving the names of columns containing URIs to read.
     :param band_indexes: list of integers indicating which bands, zero-based, to read from the raster files specified; default is to read only the first band.
     :param tile_dimensions: tuple or list of two indicating the default tile dimension as (columns, rows).
+    :param buffer_size: buffer each tile read by this many cells on all sides.
     :param lazy_tiles: If true (default) only generate minimal references to tile contents; if false, fetch tile cell values.
```
Comment (metasim) [inline pyrasterframes/src/main/python/pyrasterframes/__init__.py:139]: You rock! 🤘

File: pyrasterframes/src/main/python/pyrasterframes/__init__.py [hunk 14]
```diff
@@ -249,6 +252,24 @@ def temp_name():
...
 
+def _stac_api_reader(
+        df_reader: DataFrameReader,
+        uri: str,
+        filters: dict = None,
+        search_limit: Optional[int] = None) -> DataFrame:
+    """
+    uri - STAC API uri
+    filters - a STAC API Search filters dict (bbox, datetime, intersects, collections, items, limit, query, next)
+    search_limit - search results convenient limit method
```
Comment (metasim) [inline pyrasterframes/src/main/python/pyrasterframes/__init__.py:263]: This might be better as named parameters, or as `**kwargs`.

Comment (metasim) [reply pyrasterframes/src/main/python/pyrasterframes/__init__.py:263]: Discoverability when working in a notebook or something would benefit from these parameters being fully spelled out with defaults.

File: pyrasterframes/src/main/python/pyrasterframes/rasterfunctions.py [hunk 15]
```diff
@@ -781,6 +780,67 @@ def rf_identity(tile_col: Column_type) -> Column:
...
+
+def rf_hillshade(tile_col: Column_type, azimuth: Union[int, float, Column_type], altitude: Union[int, float, Column_type], z_factor: Union[int, float, Column_type]) -> Column:
+    """Calculates the hillshade of each cell in an elevation raster"""
+    if isinstance(azimuth, (int, float)):
+        azimuth = lit(azimuth)
+    if isinstance(altitude, (int, float)):
+        altitude = lit(altitude)
+    if isinstance(z_factor, (int, float)):
+        z_factor = lit(z_factor)
+    return _apply_column_function('rf_hillshade', tile_col, azimuth, altitude, z_factor)
```
Comment (metasim) [inline pyrasterframes/src/main/python/pyrasterframes/rasterfunctions.py:843]: Nicely done!!

Comment (metasim) [reply pyrasterframes/src/main/python/pyrasterframes/rasterfunctions.py:843]: Sorry there's documentation duplicated. In another universe I would have figured out how to have the Python API generated from the annotations on the Expressions.

File: pyrasterframes/src/main/python/pyrasterframes/rf_types.py [hunk 16]
```diff
@@ -390,6 +390,11 @@ def __init__(self, cells, cell_type=None):
                 # if the value in the array is `nd_value`, it is masked as nodata
                 self.cells = np.ma.masked_equal(self.cells, nd_value)
 
+        # is it a buffer tile? crop it on extraction to preserve the tile behavior
```
Comment (metasim) [inline pyrasterframes/src/main/python/pyrasterframes/rf_types.py:393]: Not sure this is the right thing to do, but don't want to hold up the PR for this. Back to the "everything's a buffered tile" perspective, I'd want to explore supporting buffered tile semantics on the Python mirror type.

File: pyrasterframes/src/main/python/pyrasterframes/rf_types.py [hunk 17]
```diff
@@ -533,15 +540,16 @@ def deserialize(self, datum):
...
         except ValueError as e:
             raise ValueError({
                 "cell_type": cell_type,
                 "cols": cols,
                 "rows": rows,
                 "cell_data.length": len(cell_data_bytes),
                 "cell_data.type": type(cell_data_bytes),
-                "cell_data.values": repr(cell_data_bytes)
+                "cell_data.values": repr(cell_data_bytes),
+                "grid_bounds": datum.gridBounds
```
Comment (metasim) [inline pyrasterframes/src/main/python/pyrasterframes/rf_types.py:552]: 🙏

File: rf-notebook/src/main/docker/Dockerfile [hunk 18]
```diff
@@ -1,21 +1,21 @@
 # jupyter/scipy-notebook isn't semantically versioned.
 # We pick this arbitrary one from Sept 2019 because it's what latest was on Oct 17 2019.
-FROM jupyter/scipy-notebook:7a0c7325e470
+FROM jupyter/scipy-notebook:python-3.8.8
```
Comment (metasim) [inline rf-notebook/src/main/docker/Dockerfile:3]: 👍

**Review by metasim — COMMENTED** [review]
🤘 🧙 💪 :shipit: 

Can you update the `release-notes.md` file under a `0.10` heading? I forgot to ask for that in the last PR too.

File: core/src/main/scala/org/locationtech/rasterframes/ref/RasterRef.scala (hunk 9 above)
Comment (pomadchin) [reply core/src/main/scala/org/locationtech/rasterframes/ref/RasterRef.scala:37]: @metasim I think subgrid is to read the window (that is the old API), the buffer is the size of the actual underlying this window source.

File: core/src/test/scala/org/locationtech/rasterframes/functions/FocalFunctionsSpec.scala (hunk 10 above)
Comment (pomadchin) [reply core/src/test/scala/org/locationtech/rasterframes/functions/FocalFunctionsSpec.scala:42]: @metasim 🤦 I didn't figure out how to make it better.

File: pyrasterframes/src/main/python/pyrasterframes/rf_types.py (hunk 16 above)
Comment (pomadchin) [reply pyrasterframes/src/main/python/pyrasterframes/rf_types.py:393]: @metasim yes, for now it is done to preserve the old behavior and not to break anything.

File: core/src/test/scala/org/locationtech/rasterframes/functions/FocalFunctionsSpec.scala (hunk 10 above)
Comment (metasim) [reply core/src/test/scala/org/locationtech/rasterframes/functions/FocalFunctionsSpec.scala:42]: Clever hack!

File: core/src/main/scala/geotrellis/raster/mapalgebra/focal/BufferedFocalMethods.scala [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+    self.mapTile(_.slope(cs, zFactor, bounds, target))
+
+  /**
+   * Calculates the aspect of each cell in a raster.
+   *
+   * @param   cs          cellSize of the raster
+   */
+  def aspect(cs: CellSize, bounds: Option[GridBounds[Int]] = None, target: TargetCell = TargetCell.All): BufferTile =
+    self.mapTile(_.aspect(cs, bounds, target))
+}
```
[the remark this answers is not in the document]
Comment (pomadchin) [reply core/src/main/scala/geotrellis/raster/mapalgebra/focal/BufferedFocalMethods.scala:90]: https://github.com/locationtech/geotrellis/pull/3419

**Pull request merged**